lib/bots/notifybot.ts

```diff
@@ -0,0 +1,918 @@
+			// committed code.
+			this.reportError(new NotifyBotError(NotifyErrorCodes.KeyframeNotFound,
+				`${process.env.NOTIFYBOT_NAME} failed to find Keyframe versions. The reason ` +
+				`for this is:\r\n${err.message}\r\n` +
+				'Please carry out relevant changes or alert an appropriate admin.',
+				owner, repo, 'unknown'
+			));
+		});
+	}
+
+	/**
+	 * Trace each PR for a deployed/regressed component version and notify the authors of the PR of its
+	 * state, along with following any Front conversations associated with them via issues and
+	 * commenting on them.
+	 *
+	 * @param prDetails  Details of the component being deployed/regressed.
+	 * @returns           Promise on completion of notifications.
+	 */
+	private tracePRAndNotify(prDetails: DeployedPR): Promise<void> {
+		// RegExp checking for issue cross referencing.
+		const repoDetails = this.getRepoDetails(prDetails.repo);
+		if (!repoDetails) {
+			throw new Error(`Cannot find appropriate repo/owner for ${prDetails.repo}`);
+		}
+
+		// Go through each PR associated with the version.
+		return Promise.map(prDetails.prs, (pr) => {
+			const prData = {
+				number: pr,
+				owner: repoDetails.owner,
+				repo: repoDetails.repo,
+			};
+
+			// Get the PR. Get all comments. Look for links to issues in both, including links to HQ issues.
+			return Promise.join(this.dispatchToEmitter(this.githubEmitterName, {
+					data: prData,
+					method: this.githubApi.issues.getComments,
+				}),
+				this.dispatchToEmitter(this.githubEmitterName, {
+					data: prData,
+					method: this.githubApi.pullRequests.get
+				}),
+				(comments: GithubApiTypes.IssueComment[], pullRequest: GithubApiTypes.PullRequest) => {
+					// Find reference to parent issues. We should not have any Front conversations in a PR.
+					return  _.concat(_.flatMap(comments, (comment) => this.matchIssue(comment.body, IssueRefRE)),
+						this.matchIssue(pullRequest.body, IssueRefRE),
+						_.flatMap(comments, (comment) => this.matchIssue(comment.body, HqRE)),
+						this.matchIssue(pullRequest.body, HqRE));
+				}
+			).then((issueNumbers) => {
+				// Get every issue.
+				return Promise.mapSeries(issueNumbers, (issueNumber) => {
+					let topicIssues: TopicIssues;
+
+					// Get any Front topics that exist on the issue.
+					return this.getTopicsOnIssue(issueNumber, repoDetails.owner, repoDetails.repo)
+					.then((results) => {
+						topicIssues = results;
+						return this.retrieveTopics(topicIssues.topics);
+					}).then((conversations: string[]) => {
+						// Now we have all Front threads. Examine each of them for their IDs, and then post a private
+						// comment into them to note updated references.
+						let bodyMessage: string;
+						if (prDetails.incrementing) {
+							bodyMessage = `RESOLVED: Deployed version ${prDetails.deployedVersion} of ${prDetails.repo} on ` +
+								`the ${prDetails.environment} has resolved the following issues that are attached ` +
+								'to this conversation:';
```

Comment:
This changes how we'd specced the bot though. I think we need to discuss this again and decide how we're going to determine the actual fixed Issue. From our Commit/PR spec.: 'We don't use the github fixes, closes et al. keywords, because these result in the associated issue to be closed as soon as the PR is merged.' Issues aren't closed until the PR is deployed, but NB is only going to activate when a Keyframe has just been deployed to an environment, which is the first time an Issue can actually said to be closed.

So, we have a problem. We're going to need to determine the Issue that's actually closed. We could, for example, say that the body for a PR must *only* include Issues that are resolved by this PR. That way any further comments on the PR can include cross-referenced issues that aren't actually resolved. Subsequently, Issues for a PR that cross-reference an HQ Issue should only connect them in the body? (This is actually slightly mitigated because NB will only follow HQ issues referenced by `hq:`). Or we could use a separate keyword that won't close the Issue but isn't one we use for anything else.

cc @lekkas